### PR TITLE
Add route factory and onion transport

### PR DESCRIPTION
**Pull request by david415**
- Add route factory for building routes for clients
- Add tor onion transport
- Add integration tests using travis + tox + chutney + tor
  - Bob send Alice a message and it gets routed through the mix network composed of onion services
- many bug fixes
TODO:
- add eliot logging

### CODE REVIEW

File: txmix/client.py [hunk 1]
```diff
@@ -60,10 +62,33 @@ def send(self, route, message):
...
+@implementer(IRouteFactory)
+@attr.s
+class RandomRouteFactory(object):
+    """
+    I create random routes.
+    """
+    params = attr.ib(validator=attr.validators.instance_of(SphinxParams))
+    pki = attr.ib(validator=attr.validators.provides(IMixPKI))
+    rand_reader = attr.ib(validator=attr.validators.provides(IReader))
+
```
Comment (meejah) [inline txmix/client.py:74]: its not clear why this is called rand_reader (not e.g. just "reader") nor what it does to me; maybe a comment?

File: txmix/onion_transport.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            protocol.transport.write(message)
+            protocol.loseConnection()
+
+        d.addCallback(is_connected)
+        return d
+
+    # Protocol parent method overwriting
+
+    def dataReceived(self, data):
+        assert len(data) == self.sphinx_packet_size
```
Comment (meejah) [inline txmix/onion_transport.py]: This should probably buffer instead right? i.e. you could (theoretically) receive things 1 byte at a time...

File: txmix/onion_transport.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        self.protocol.received(data)
+
+    def connectionLost(self):
+        """
+        Called when the connection is shut down.
+        """
+        # XXX raise some kind of exception?
+        pass
```
Comment (meejah) [inline txmix/onion_transport.py]: You can receive "reason" arg, and see if it's `ConnectionDone` (clean loss) or something else (which will be the error).

**Comment by meejah** [comment]
Otherwise, looks good (although my overall understanding of "txmix" is still low, mind)...

File: txmix/client.py (hunk 1 above)
Comment (david415) [reply txmix/client.py:74]: it supplies random bytes from an entropy source so that key material can be generated.

Comment (david415) [reply txmix/client.py:74]: i have an explicit interface so i can mock it out in the tests. and so i can make the tests deterministic.

i should improve the tests; i've been pasting big ass strings of hex... but really i should switch to use chacha20 with a shared key.

File: txmix/onion_transport.py (hunk 2 above)
Comment (david415) [reply txmix/onion_transport.py]: yeah

File: txmix/client.py (hunk 1 above)
Comment (david415) [reply txmix/client.py:74]: maybe i should call it entropy_reader?

File: txmix/onion_transport.py (hunk 3 above)
Comment (david415) [reply txmix/onion_transport.py]: ok doing that now.

**Pull request merged**